Source/Metal/CommandBufferMTL.mm

```diff
@@ -0,0 +1,620 @@
+        BufferMTL* mtlBuffer = (BufferMTL*)buffers[i];
+        
+        const size_t slotIndex = i + baseSlot;
+        m_vertexBuffers[slotIndex].m_Offset = offsets[i];
+        m_vertexBuffers[slotIndex].m_Buffer = mtlBuffer;
+        m_dirtyVertexBufferBits |= (1 << slotIndex);
+        
+        if(m_RendererEncoder)
+            [m_RendererEncoder setVertexBuffer: mtlBuffer->GetHandle()
+                                    offset: offsets[i]
+                                   atIndex: i + baseSlot];
+    }
+    
+}
+
+void CommandBufferMTL::Draw(const DrawDesc& drawDesc) {
+    [m_RendererEncoder drawPrimitives: m_CurrentPipeline->GetPrimitiveType()
+                          vertexStart:drawDesc.baseVertex
+                          vertexCount:drawDesc.vertexNum
+                        instanceCount:drawDesc.instanceNum
+                         baseInstance: drawDesc.baseInstance];
+    
+}
+
+void CommandBufferMTL::DrawIndexed(const DrawIndexedDesc& drawIndexedDesc) {
+    id<MTLBuffer> indexBuffer = m_indexBuffer.m_Buffer->GetHandle();
+    [m_RendererEncoder drawIndexedPrimitives: m_CurrentPipeline->GetPrimitiveType()
+                              indexCount: drawIndexedDesc.indexNum
+                               indexType: m_indexBuffer.m_Type
+                             indexBuffer: indexBuffer
+                       indexBufferOffset: m_indexBuffer.m_Offset];
+}
+
+void CommandBufferMTL::DrawIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {
+    // TODO: implement count Buffer
+    NSCAssert(!countBuffer, @"count buffer not supported");
+    
+    [m_RendererEncoder
+         drawPrimitives: m_CurrentPipeline->GetPrimitiveType()
+         indirectBuffer:((BufferMTL&)buffer).GetHandle()
+         indirectBufferOffset: offset];
+}
+void CommandBufferMTL::DrawIndexedIndirect(const Buffer& buffer, uint64_t offset, uint32_t drawNum, uint32_t stride, const Buffer* countBuffer, uint64_t countBufferOffset) {
+    InsertBarriers();
+    
+    id<MTLBuffer> indexBuffer = m_indexBuffer.m_Buffer->GetHandle();
+    const BufferMTL& bufferImpl = (const BufferMTL&)buffer;
+    [m_RendererEncoder drawIndexedPrimitives: m_CurrentPipeline->GetPrimitiveType()
+                              indexCount: drawNum
+                               indexType: m_indexBuffer.m_Type
+                             indexBuffer: indexBuffer
+                       indexBufferOffset: m_indexBuffer.m_Offset
+                      ];
+}
+void CommandBufferMTL::Dispatch(const DispatchDesc& dispatchDesc) {
+    InsertBarriers();
+    
+    MTLSize threadgroupCount = MTLSizeMake(dispatchDesc.x, dispatchDesc.y, dispatchDesc.z);
+//    MTLSize threadPerThreadGroup = MTLSizeMake(dispatchDesc.x, dispatchDesc.y, dispatchDesc.z);
+//    [m_ComputeEncoder
+//     dispatchThreadgroups: threadgroupCount
+//     threadsPerThreadgroup: threadPerThreadGroup];
```

Comment:
you can pass the number of `dispatchThreadgroups` and `threadsPerThreadgroup`

Reply:
any clue how this would be managed?

Reply:
MoltenVK pulls `threadsPerThreadgroup` from shaders and stores in pipelines. This value becomes available when a pipeline gets bound.